### PR TITLE
feat: create class for config and use it

### PR DESCRIPTION
**Pull request by olgam4**
Trying to fix the TODO which states that Config should probably be a class, which I agree with. Then, the thing is accessing that object OR making is "static". What do you guys think?

### CODE REVIEW

**Comment by ThePrimeagen** [comment]
I cannot thank you enough for trying to help improve this library. This was a huge thing that I needed to do for the next refactor. Once you feel like it's ready I'll be glad to review it all.

**Comment by olgam4** [comment]
As of right now, I seem to be only missing how you spec the 106, but it should be working. Won't be long :)

**Comment by ThePrimeagen** [comment]
I do not have the instructions on running the test. I only have a very simple test for extract function. If you're familiar with plenary, you can run tests using their commands. 

Perhaps I can convince @tjdevries to create a PR to run tests on every pr.

**Comment by olgam4** [comment]
And voilà. First time collaborating on an open source project; let me know what to improve!

**Comment by ThePrimeagen** [comment]
Super excited! I'll review tonight!

**Comment by ThePrimeagen** [comment]
When you get the chance, please squash :)

File: lua/refactoring/config.lua [hunk 1]
```diff
@@ -1,92 +1,89 @@
-local config = {
```
Comment (ThePrimeagen) [inline lua/refactoring/config.lua:1]: please run stylua on this file :)

File: lua/refactoring/get_input.lua [hunk 2]
```diff
@@ -1,21 +1,14 @@
-local Config = require("refactoring.config")
-
 -- This will be able to be hot swapped out with better input capture as we
 -- go on.  This is just a place holder
-local function get_input(question, text)
+local function get_input(question, text, options)
```
Comment (ThePrimeagen) [inline lua/refactoring/get_input.lua:3]: stylua please :)

File: lua/refactoring/init.lua [hunk 3]
```diff
@@ -2,12 +2,14 @@ local refactors = require("refactoring.refactor")
 local Config = require("refactoring.config")
 
 local M = {}
+local config = {}
 
 function M.setup(config)
-    Config.setup(config)
+    config = Config.setup(config)
```
Comment (ThePrimeagen) [inline lua/refactoring/init.lua:8]: we don't use the `config` parameter within the config.setup, but that is fine, but that is fine.

File: lua/refactoring/refactor/106.lua [hunk 4]
```diff
@@ -9,19 +9,19 @@ local helpers = require("refactoring.helpers")
 local get_selected_local_defs = require(
     "refactoring.pipeline.get_selected_local_defs"
 )
-local Config = require("refactoring.config")
 
 local M = {}
 
 local function get_code(
+    options,
```
Comment (ThePrimeagen) [inline lua/refactoring/refactor/106.lua:16]: we could collapse options, bufnr, and lang down to just `refactor` object.

Side note, at some point I will also be refactoring the refactor object into a class of some sort.

File: lua/refactoring/refactor/106.lua [hunk 5]
```diff
@@ -9,19 +9,19 @@ local helpers = require("refactoring.helpers")
 local get_selected_local_defs = require(
     "refactoring.pipeline.get_selected_local_defs"
 )
-local Config = require("refactoring.config")
```
Comment (ThePrimeagen) [inline lua/refactoring/refactor/106.lua:12]: stylua please :)

File: lua/refactoring/tests/refactor_spec.lua [hunk 6]
```diff
@@ -1,7 +1,7 @@
 local scandir = require("plenary.scandir")
 local Path = require("plenary.path")
 local refactoring = require("refactoring.refactor")
-local Config = require("refactoring.config")
+local Config = require('refactoring.config')
```
Comment (ThePrimeagen) [inline lua/refactoring/tests/refactor_spec.lua:4]: stylua please :)

File: lua/refactoring/tests/refactor_spec.lua [hunk 7]
```diff
@@ -94,13 +94,14 @@ describe("Refactoring", function()
                 )
             )
             vim.api.nvim_buf_set_lines(0, 0, -1, false, contents)
-            Config.automate_input(inputs)
+	    local options = Config.setup()
```
Comment (ThePrimeagen) [inline lua/refactoring/tests/refactor_spec.lua:97]: stylua please :)

You can also pull this out to the next scope.

File: lua/refactoring/config.lua [hunk 8]
```diff
@@ -1,92 +1,89 @@
...
+    conf.formatting = default_formatting
+    conf.code_generation = default_code_generation
+    return conf
+end
 
-function M.get_config()
-    return config
+function Config:automate_input(inputs)
+    if type(inputs) ~= "table" then inputs = {inputs} end
+    self._reset_inputs(inputs)
```
Comment (ThePrimeagen) [inline lua/refactoring/config.lua:67]: I believe you will lose `self` here.

Make sure you use `self:_reset_inputs(inputs)`

File: lua/refactoring/config.lua [hunk 9]
```diff
@@ -1,92 +1,89 @@
...
+function Config:get_code_generation_for(lang) return self.code_generation[lang] end
+
+function Config:get_formatting_for(filetype) return self.formatting[filetype] end
 
-    -- TODO: This feels wrong.  I feel like config needs to become an object
-    -- TODO: Refactor this into a class.
-    config._automation.inputs = inputs
-    config._automation.inputs_idx = 0
+function Config:get_next_input()
+    self._automation.current_input_index = #self._automation.current_input_index + 1
```
Comment (ThePrimeagen) [inline lua/refactoring/config.lua:80]:
```lua
self._automation.current_input_index = #self._automation.current_input_index + 1
-----------------oops------------------^ // remove #
```

**Review by ThePrimeagen — CHANGES_REQUESTED** [review]
Small feedback.

Let me know once that is done :)

File: lua/refactoring/init.lua (hunk 3 above)
Comment (olgam4) [reply lua/refactoring/init.lua:8]: I know, but it's marked as a TODO 🤷‍♀️

File: lua/refactoring/refactor/106.lua (hunk 4 above)
Comment (olgam4) [reply lua/refactoring/refactor/106.lua:16]: I agree. We could open an issue and I can start working on that too, if you want!

**Comment by olgam4** [comment]
Screen Shot 2021-08-04 at 8 35 17 AM

Every commit has been squashed and fixes requested are done!

**Comment by ThePrimeagen** [comment]
It appears you haven't run stylua yet!  Perhaps you are not pointing it at the toml provided with this project

**Comment by olgam4** [comment]
So as said on the Discord, I'm having real problems running the tests, but the feature works on my end. I would like help on the tests if possible, or I can put this one on hold until I fix them alone.

File: lua/refactoring/init.lua [hunk 10]
```diff
@@ -20,11 +24,14 @@ function M.refactor(name)
...
     -- Remove the calls to vim.fn
     -- I just forgot the name of this ;)
-    refactors[refactor](vim.api.nvim_buf_get_number(0))
+    refactors[refactor](vim.api.nvim_buf_get_number(0), options)
 end
 
 function M.get_refactors()
     return vim.tbl_keys(refactors.refactor_names)
 end
+GLOBAL = {}
```
Comment (ThePrimeagen) [inline lua/refactoring/init.lua:33]: You have a global config!

**Comment by ThePrimeagen** [comment]
so you have accidentally rebased my commits into your branch!  Ohh nos!  Very sorry you got yourself into this, but hey! Its a learning experience!

**Review by ThePrimeagen — COMMENTED** [review]
> so you have accidentally rebased my commits into your branch! Ohh nos! Very sorry you got yourself into this, but hey! Its a learning experience!

And you fixed it!  Well done :)

**Comment by olgam4** [comment]
Yeah I figured I messed something up ahaha but everything is back to normal now, I still confilct :)

File: lua/refactoring/tests/extract.simple-function.expected.lua [hunk 11]
```diff
@@ -1,14 +1,13 @@
-
 local function foo_bar(a, test, test_other)
-        for idx = test - 1, test_other do
+    for idx = test - 1, test_other do
```
Comment (ThePrimeagen) [inline lua/refactoring/tests/extract.simple-function.expected.lua:2]: next time avoid doing to much of a refactor.  these could of remained, make a ticket, and we could of resolved it shortly thereafter.

Comment (olgam4) [reply lua/refactoring/tests/extract.simple-function.expected.lua:2]: It is due to formatting, I havent touched those files

File: lua/refactoring/tests/refactor_spec.lua [hunk 12]
```diff
@@ -1,7 +1,7 @@
 local scandir = require("plenary.scandir")
 local Path = require("plenary.path")
-local refactoring = require("refactoring.refactor")
-local Config = require("refactoring.config")
+local refactoring = require("lua.refactoring.refactor.init")
```
Comment (ThePrimeagen) [inline lua/refactoring/tests/refactor_spec.lua:3]: this file really needs stylua ran on it

**Review by ThePrimeagen — CHANGES_REQUESTED** [review]

**Comment by ThePrimeagen** [comment]
Could we jump on a discord call?

**Comment by olgam4** [comment]
> Could we jump on a discord call?

yep! is now a good time? waiting on the discord :)

**Pull request closed without merging**